gtsam/navigation/ConstantVelocityFactor.h

```diff
@@ -0,0 +1,68 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    ConstantVelocityFactor.h
+ * @brief   Maintain a constant velocity motion model between two NavStates
+ * @author  Asa Hammond
+ */
+
+#include <gtsam/navigation/NavState.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+namespace gtsam {
+
+/**
+ * Binary factor for applying a constant velocity model to a moving body represented as a NavState.
+ * The only measurement is dt, the time delta between the states.
+ */
+class ConstantVelocityFactor : public NoiseModelFactor2<NavState, NavState> {
+    double dt_;
+
+   public:
+    using Base = NoiseModelFactor2<NavState, NavState>;
+
+   public:
+    ConstantVelocityFactor(Key i, Key j, double dt, const SharedNoiseModel &model)
+        : NoiseModelFactor2<NavState, NavState>(model, i, j), dt_(dt) {}
+    ~ConstantVelocityFactor() override{};
+
+    /**
+     * @brief Caclulate error: (x2 - x1.update(dt)))
+     * where X1 and X1 are NavStates and dt is
+     * the time difference in seconds between the states.
+     * @param x1 NavState for key a
+     * @param x2 NavState for key b
+     * @param H1 optional jacobian in x1
+     * @param H2 optional jacobian in x2
+     * @return * Vector
+     */
+    gtsam::Vector evaluateError(const NavState &x1, const NavState &x2,
+                                boost::optional<gtsam::Matrix &> H1 = boost::none,
+                                boost::optional<gtsam::Matrix &> H2 = boost::none) const override {
+        // only used to use update() below
+        static const Vector3 b_accel{0.0, 0.0, 0.0};
+        static const Vector3 b_omega{0.0, 0.0, 0.0};
+
+        Matrix predicted_H_x1;
+        NavState predicted = x1.update(b_accel, b_omega, dt_, predicted_H_x1, {}, {});
```

Comment:
Cool. But please use `H1 ? &predicted_H_x1  : nullptr` to avoid evaluation of derivative if H1 is not asked for.

Reply:
I couldn't get this style to compile, so I went a little more verbose. Any thoughts on how to make this prettier would be appreciated, otherwise this accomplishes your request I believe.

Reply:
```error: no viable conversion from 'gtsam::Matrix *' (aka 'Matrix<double, Dynamic, Dynamic> *') to 'OptionalJacobian<9, 9>```

Reply:
Ah, yeah. This style should definitely compile, and I recommend you do use it. But the issue is you need a *fixed size* matrix. Which will also benefit performance, as it gets rid of a heap allocation. In example above:
```
Matrix99 predicted_H_x1;
```

Reply:
Ahh... got it. That compiles now. Should be done!